### PR TITLE
perf(encoding): #18 Phase 5 — LDM producer (gear hash + bucket table + search/emit)

### LINKED ISSUE #18: feat: long distance matching (LDM) for compression
## Summary

Long Distance Matching enables finding matches beyond the normal window size, critical for highly repetitive data with distant patterns (log files, database dumps, backup archives).

### Gear hash rolling
- Rolling hash: `hash = (hash << 1) + ZSTD_ldm_gearTab[byte]`
- Gear table: 256-entry lookup table for byte-to-hash-value mapping
- Stop mask determines split points (average every `2^hashRateLog` bytes)

### LDM hash table
- Bucket-based: `ldmEntry_t` with position + checksum
- `LDM_BUCKET_SIZE_LOG = 4` (16 entries per bucket)
- Memory-efficient: only stores position + small checksum, not full content

### Parameters
- `ZSTD_c_enableLongDistanceMatching`: on/off
- `ldmHashLog`: Hash table size
- `ldmMinMatch`: Minimum match length (default 64 bytes)
- `ldmBucketSizeLog`: Collision bucket size
- `ldmHashRateLog`: Insertion frequency

### Integration
- LDM matches are merged with regular matches in sequence generation
- Auto-enables large window (128MB) for effective long-range matching

## What needs to be implemented

1. **Gear hash implementation** — rolling hash with lookup table
2. **LDM hash table** — bucket-based with position + checksum
3. **LDM match finding** — search LDM table, verify matches, extend
4. **Sequence merging** — combine LDM matches with regular matches
5. **Parameter API** — enable/configure LDM via compression parameters
6. **Window extension** — auto-increase window when LDM enabled

## Acceptance criteria
- [ ] LDM can be enabled via parameter
- [ ] Finds matches beyond normal window distance
- [ ] Decompressible by C zstd
- [ ] Compression ratio improves on highly repetitive large data
- [ ] Configurable parameters

## Time estimate
3d

### PR DESCRIPTION
**Pull request by polaz**
## Summary

Phase 5 of [#111](https://github.com/structured-world/structured-zstd/issues/111) — implements [#18](https://github.com/structured-world/structured-zstd/issues/18) (long-distance matching). Lands the full donor LDM producer (gear rolling hash + bucket-based hash table + verify/extend search + aggregate driver) under a new `encoding/ldm/` module. Two commits, ~2.1k LoC including 33 unit tests.

## Activation policy (distro-grade)

**`LdmProducer` is never activated automatically by any `CompressionLevel` preset.** This mirrors upstream `libzstd.so.1` behaviour: `ZSTD_compress(..., level)` keeps LDM off at every standard level (1..22). The opt-in surface lands separately:

| Surface | LDM activation | Status |
|---|---|---|
| `compress(level)` / `CompressionLevel::*` | **off**, every level | preserved by this PR |
| Rust parameter API (`set_parameter(EnableLdm, true)`) | **on** | tracked in [#27](https://github.com/structured-world/structured-zstd/issues/27) |
| C ABI (`ZSTD_c_enableLongDistanceMatching`) | **on** | tracked in #126 / #127 |
| `zstd --long[=N]` CLI flag | **on** | tracked in #128 |

`BtMatcher::ldm_producer: Option<LdmProducer>` stays `None` until those issues land — preserves byte-parity with upstream `ZSTD_compress(..., 22)` and keeps the `level22_sequences_match_donor_on_corpus_proxy` ratio gate green.

## Module structure

```
zstd/src/encoding/ldm/
├── mod.rs        — LdmProducer aggregator + generate_into pipeline
├── gear_hash.rs  — ZSTD_ldm_gear_init/reset/feed + verbatim 256-entry GEAR_TAB
├── params.rs     — LdmParams + ZSTD_ldm_adjustParameters port
├── table.rs      — LdmHashTable (round-robin bucket insert)
└── search.rs     — count_backwards_match + find_best_match (prefix-only path)
```

## Donor parity anchors

Every primitive cites \`lib/compress/zstd_ldm.c\` v1.5.7 by file:line:

* \`gear_hash::GEAR_TAB\` ← \`zstd_ldm_geartab.h\` (verbatim, 256-entry permutation table, cross-checked zero-mismatch against parsed donor header)
* \`gear_hash::GearHashState::new\` ← \`ZSTD_ldm_gear_init\` (zstd_ldm.c:32)
* \`gear_hash::reset\` ← \`ZSTD_ldm_gear_reset\` (zstd_ldm.c:65)
* \`gear_hash::feed\` ← \`ZSTD_ldm_gear_feed\` (zstd_ldm.c:96)
* \`params::LdmParams::adjust_for\` ← \`ZSTD_ldm_adjustParameters\` (zstd_ldm.c:135)
* \`table::LdmHashTable::insert\` ← \`ZSTD_ldm_insertEntry\` (zstd_ldm.c:198)
* \`search::count_backwards_match\` ← \`ZSTD_ldm_countBackwardsMatch\` (zstd_ldm.c:214)
* \`search::find_best_match\` ← per-split inner loop (zstd_ldm.c:405-466)
* \`LdmProducer::generate_into\` ← \`ZSTD_ldm_generateSequences_internal\` (zstd_ldm.c:346-515) prefix-only path

## Implementation status

* **Prefix-only path** — fully implemented. Equivalent to upstream on every input that fits in a single contiguous history slice (the only path the current Rust encoder exercises — no separate \`extDict\` buffer surface).
* **Two-segment (extDict) backward path** — documented and deferred. \`search.rs\` cites the donor functions (\`ZSTD_count_2segments\`, \`ZSTD_ldm_countBackwardsMatch_2segments\`) but does not yet implement them; lands together with the eventual dictionary-history support work.

## Test coverage

33 new LDM unit tests + 1 end-to-end producer test:

| Module | Tests | Coverage |
|---|---|---|
| \`gear_hash\` | 9 | GEAR_TAB anchor entries (donor cross-check), stop_mask derivation (3 paths), recurrence regression, reset/feed equivalence, zero-mask split-per-byte, concatenation invariant, batch-size pre-condition |
| \`params\` | 5 | strategy→hash_rate_log table, hash_log clamp, min_match halving at btultra, bucket clamp, derived helpers |
| \`table\` | 7 | sizing math, bucket clamp, round-robin wrap, adjacent-bucket isolation, clear+rewind, donor-max hash_log, bucket_mask |
| \`search\` | 9 | count_backwards_match (mismatch stop + anchor bound), find_best_match (stale rejection, checksum filter, longest-combined selection, backward extension, short-forward rejection) |
| \`mod\` (aggregate) | 3 | producer construction with btultra2 defaults, clear-resets-rolling-hash, empty-range no-op |
| \`mod\` (e2e) | 1 | 4 KiB payload + 64 KiB gap + 4 KiB repeat → producer emits at least one HcRawSeq crossing the gap |

## Build + suite

* **460 / 460** lib tests pass on the full suite (was 427 pre-Phase-5 → +33 LDM tests).
* \`cargo clippy --lib --tests\` clean (no warnings).
* \`level22_sequences_match_donor_on_corpus_proxy\` ratio gate **PASS**.
* No new dependencies; reuses existing \`twox-hash::XxHash64\` already pulled in for frame checksums.

## Commits

* \`8cd597b\` — Phase 5 foundations: gear hash + params + bucket table + LdmProducer construction + fill path (gear walk + XXH64 + bucket insert)
* \`7759eef\` — Phase 5 producer: search.rs (count_backwards_match + find_best_match) + generate_into rewritten end-to-end to emit HcRawSeq entries via verify/extend pipeline

## Test plan

- [x] \`cargo nextest run --lib\` — 460/460 green
- [x] \`cargo clippy --lib --tests\` — clean
- [x] \`level22_sequences_match_donor_on_corpus_proxy\` — PASS (byte-parity preserved, LDM stays None by default)
- [x] End-to-end producer test — long-range match emitted, offset crosses the engineered gap
- [ ] CI matrix (x86_64-linux, aarch64-linux, x86_64-darwin, aarch64-darwin, i686-linux) — pending GitHub Actions
- [ ] Codecov patch coverage — pending

## Related issues

* Closes [#18](https://github.com/structured-world/structured-zstd/issues/18) — long-distance matching
* Part of [#111](https://github.com/structured-world/structured-zstd/issues/111) Phase 5
* Unblocks [#27](https://github.com/structured-world/structured-zstd/issues/27) — configurable parameter API (LDM activation hook will plug into \`BtMatcher::ldm_producer\`)

Closes #18

## Summary by CodeRabbit

* **New Features**
  * Feature-gated long-distance-match (LDM) encoder: gear-hash split detection, bucketed LDM table, candidate search, and a producer pipeline; BT matcher now uses live-history context for LDM candidates.
* **Bug Fixes**
  * Prevented out-of-bounds panics by validating absolute-coordinate relationships and clearing producer state on reset.
* **Tests**
  * Extensive unit and regression tests covering hashing, parameter derivation, table behavior, candidate search, producer flow, and panic prevention.
* **Documentation**
  * Rewrote README and crate/module docs; enabled richer feature-gated docs for published builds.

[Review Change Stack](https://app.coderabbit.ai/change-stack/structured-world/structured-zstd/pull/139?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: zstd/src/encoding/bt/mod.rs [hunk 1]
```diff
@@ -305,15 +316,51 @@ impl BtMatcher {
...
+        if let Some(producer) = self.ldm_producer.as_mut() {
+            // Translate absolute → slice index. The caller is
+            // responsible for ensuring `current_abs_start >=
+            // history_abs_start` (the optimal driver derives both
+            // from the same `MatchTable`, so the invariant holds
+            // structurally).
+            debug_assert!(current_abs_start >= history_abs_start);
+            let block_start = current_abs_start - history_abs_start;
+            let block_end = block_start.saturating_add(current_len).min(history.len());
+            producer.generate_into(history, block_start, block_end, &mut self.ldm_sequences);
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/bt/mod.rs]: @coderabbitai Fixed in commit b036a1d. The producer now stores absolute stream positions in every cross-block surface:

- `LdmEntry.offset` is absolute (donor `base`-relative)
- `generate_into` takes `(live_history, history_abs_start, block_start_abs, block_end_abs, out)` and translates abs→slice only at the moment of `live_history[..]` access
- `find_best_match` takes absolute coordinates + `history_abs_start`; the donor `entry.offset <= lowestIndex` staleness check (with `lowest_index_abs = history_abs_start`) filters entries that fall out of the window after a slide
- Call site in `match_generator.rs` now passes `self.table.live_history()` (the post-`history_start` slice) instead of the full `history` Vec, so `live_history[0]` aligns with `history_abs_start`

New regression test `generate_into_preserves_bucket_entries_across_history_slide`: 4 KiB payload + 32 KiB gap + 4 KiB repeat, call 1 with `history_abs_start = 0` populates the bucket with absolute offsets inside payload #1, call 2 with `history_abs_start = 2048` (simulated eviction) hits the long-range match into payload #1’s surviving tail. Pre-fix slice-relative offsets would either miss the match or emit a back-ref > live history length.

File: zstd/src/encoding/ldm/gear_hash.rs [hunk 2]
```diff
@@ -0,0 +1,531 @@
...
+            // without trying to bias toward `min_match_length` bits.
+            // Donor `zstd_ldm.c:56`.
+            (1u64 << hash_rate_log) - 1
+        };
+        Self {
+            rolling: GEAR_HASH_INIT,
+            stop_mask,
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/ldm/gear_hash.rs:128]: @coderabbitai Applied — `GearHashState::new` now does `hash_rate_log.min(63)` before the shift (commit 70aca66, `gear_hash.rs:108`). All production callers stay at 4..7 via `LdmParams::adjust_for`, but the clamp keeps the `1u64 << hash_rate_log` shift safe for future callers reaching for the `pub(crate)` surface directly.

File: zstd/src/encoding/ldm/search.rs [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+/// Returns the number of matched backward bytes (caps at the
+/// tighter of the two `min(p_in_abs - anchor, p_match_abs -
+/// match_base)` distances).
+pub(crate) fn count_backwards_match(
+    history: &[u8],
+    p_in_abs: usize,
+    anchor_abs: usize,
+    p_match_abs: usize,
+    match_base_abs: usize,
+) -> usize {
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/ldm/search.rs:74]: @coderabbitai Fixed — renamed `count_backwards_match` parameters to `*_idx` and updated the docstring to spell out that callers must translate `abs - history_abs_start` to slice indices before invoking (commit 70aca66). Same rename for the abs/idx terminology mismatch raised in the inline thread.

File: zstd/src/encoding/ldm/table.rs [hunk 4]
```diff
@@ -0,0 +1,579 @@
...
+
+    /// Reset every bucket to "empty" without reallocating. Donor
+    /// equivalent is the `ZSTD_cwksp` clear of the LDM region at
+    /// frame boundaries. `LdmEntry` is `Copy` so the slice
+    /// `fill` compiles down to a bulk memset — meaningful when
+    /// `hash_log` is large and this runs at every frame boundary.
+    pub(crate) fn clear(&mut self) {
+        self.entries.fill(LdmEntry::default());
+        self.bucket_offsets.fill(0);
+    }
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/ldm/table.rs:178]: @coderabbitai Fixed in commit c926dad — `clear()` now resets `position_base = 0` so a fresh frame can insert at absolute 0 without tripping the `abs_pos >= position_base` assertion. New `clear_resets_position_base` regression verifies the round-trip: insert → reduce → clear → insert-at-0.

File: zstd/src/encoding/ldm/table.rs [hunk 5]
```diff
@@ -0,0 +1,608 @@
...
+                checksum,
+            },
+        );
+    }
+
+    /// Ensure that absolute position `abs_pos` can be stored as a
+    /// `u32` offset relative to [`Self::position_base`] without
+    /// overflow. If the relative position would exceed `u32::MAX
+    /// - REBASE_GUARD_BAND`, advance `position_base` by
+    /// [`REBASE_GUARD_BAND`] and run [`Self::reduce`] across all
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/ldm/table.rs]: @coderabbitai Already fixed locally in a pending commit (race with the CI failure that surfaced this) — pushing momentarily. The line was rephrased to `When the relative position would exceed `u32::MAX − REBASE_GUARD_BAND``, no leading dash, which clippy 1.95 accepts under `-D warnings`.

File: zstd/src/encoding/ldm/table.rs [hunk 6]
```diff
@@ -0,0 +1,608 @@
...
+    pub(crate) fn insert_absolute(&mut self, hash_id: u32, abs_pos: usize, checksum: u32) {
+        debug_assert!(abs_pos >= self.position_base);
+        let rel = abs_pos - self.position_base;
+        debug_assert!(
+            rel < u32::MAX as usize,
+            "insert position {abs_pos} (rel {rel}) exceeds u32 window; \
+             producer must call `ensure_room_for` before insert"
+        );
+        // +1 bias so 0 stays reserved for the empty-slot sentinel.
+        let stored = (rel + 1) as u32;
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/ldm/table.rs]: @coderabbitai Applied in commit 0475405: `abs_pos.checked_sub(position_base).unwrap_or_else(|| panic!(..))` for the lower bound, `assert!(rel < u32::MAX as usize)` for the upper, and the `+1` bias now lives inside the `u32` arithmetic so the intermediate `usize` add cannot overflow on 32-bit. New `insert_absolute_panics_below_position_base` regression covers the underflow guard with `#[should_panic]`.

**Comment by polaz** [comment]
@coderabbitai Re: the `lowest_index_abs` exclusive-bound finding — addressed in commit e1932b2, but with a slightly different shape than the suggested `history_abs_start.saturating_sub(1)`. Instead of keeping the donor `<=` semantics and biasing the caller, I flipped the filter to inclusive `match_abs < lowest_index_abs` (i.e. `match_abs >= lowest_index_abs` survives) and kept the caller passing `history_abs_start` directly. Same effect on the edge case (`live_history[0]` remains matchable), no `saturating_sub` boundary special-case at frame start, and the field docstring spells out the donor deviation. Test fixture for `find_best_match_rejects_stale_entries` adjusted to `lowest_index_abs = 5` so the floor still rejects offset 4.

**Comment by polaz** [comment]
@coderabbitai Re: the `lowest_index_abs` exclusive-bound finding re-raised in the latest review body — it is **stale**, addressed in commit e1932b2 (round 8). Proof from the current code on `feat/#18-phase5-ldm @ a66e92f`:

`zstd/src/encoding/ldm/search.rs:219`:
```rust
if match_abs < lowest_index_abs {
    continue;
}
```

The filter is now `match_abs < lowest_index_abs` (**inclusive lower bound** — `match_abs == lowest_index_abs` survives). The caller in `LdmProducer::generate_into` (mod.rs:332) passes `history_abs_start` directly, so an entry at the very left edge of the live window (`live_history[0]`, absolute `history_abs_start`) remains matchable. No `saturating_sub` needed. Field docstring at search.rs:122-141 documents the inclusive semantics plus the donor deviation note.

**Comment by polaz** [comment]
@coderabbitai Re: the re-raised `lowest_index_abs` exclusive-bound duplicate — still **stale** after re-verifying against current `feat/#18-phase5-ldm @ 1909218`. The filter at `search.rs:219` is `if match_abs < lowest_index_abs { continue }` (inclusive lower bound). Caller in `mod.rs:332` passes `history_abs_start` directly, edge entry at `live_history[0]` survives. Field docstring at `search.rs:122-141` documents the inclusive semantics with explicit donor-deviation note. Resolution and code reasoning recorded in round 8 commit e1932b2 and round 9 PR comment #4461263264. Please re-evaluate against the current diff rather than the round-7 snapshot.

File: zstd/src/encoding/ldm/table.rs [hunk 7]
```diff
@@ -0,0 +1,638 @@
...
+        // Leave at least `REBASE_GUARD_BAND` of u32 headroom
+        // above the current insert position so the next batch of
+        // inserts doesn't immediately re-rebase.
+        let max_rel = u32::MAX as usize - REBASE_GUARD_BAND as usize;
+        if rel > max_rel {
+            // Shift the base forward by `REBASE_GUARD_BAND` —
+            // matches donor's "subtract the reducer value, clamp
+            // anything below to 0" semantics.
+            self.reduce(REBASE_GUARD_BAND);
+        }
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/ldm/table.rs:365]: @coderabbitai Fixed in 8be94ad — `ensure_room_for` now loops with `while abs_pos - position_base > max_rel { reduce(REBASE_GUARD_BAND) }`, and a regression test (`ensure_room_for_loops_across_multiple_guard_bands`) covers a 5×guard-band jump that the old single-shot `if` would have left in the panic zone.

File: zstd/src/encoding/ldm/table.rs [hunk 8]
```diff
@@ -0,0 +1,678 @@
...
+    /// bucket for `hash_id`. The caller is responsible for
+    /// invoking [`Self::ensure_room_for`] beforehand so the
+    /// relative-offset arithmetic stays inside `u32`.
+    ///
+    /// # Panics
+    ///
+    /// Panics in debug builds if `abs_pos < position_base` or if
+    /// `abs_pos - position_base + 1 > u32::MAX`. The producer
+    /// must call `ensure_room_for` before the first insert that
+    /// could exceed the window.
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/ldm/table.rs]: @coderabbitai Fixed in 23d5474 — rustdoc now states the panics fire in all build modes and cites the underlying `checked_sub(...).unwrap_or_else(panic!)` and `assert!` so the contract cannot be silently regressed to `debug_assert!`.

File: zstd/src/encoding/ldm/table.rs [hunk 9]
```diff
@@ -0,0 +1,682 @@
...
+        let bucket_start = (hash_id as usize) << self.effective_bucket_log;
+        let offset = self.bucket_offsets[hash_id as usize] as usize;
+        self.entries[bucket_start + offset] = entry;
+        // Post-increment modulo bucket size. Donor stores the result
+        // in a BYTE (so the mask is implicit at bucket_size_log <= 8);
+        // we mirror the explicit mask so values above 8 (clamped by
+        // params, but defensive here) still wrap correctly.
+        let next = (offset + 1) & slot_mask;
+        self.bucket_offsets[hash_id as usize] = next as u8;
+    }
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/ldm/table.rs:253]: @coderabbitai Fixed in c1f5b44 — `insert()` now `assert!(entry.offset != 0)` (runtime, not `debug_assert!`), the round-robin fixture shifted to offsets `1..=6`, and a regression test (`insert_panics_on_sentinel_offset_zero`) pins the new contract.

File: zstd/src/encoding/mod.rs [hunk 10]
```diff
@@ -1,4 +1,37 @@
...
+//!   `input_size + output_size`; prefer [`compress`] or
+//!   [`StreamingEncoder`] when the input is large or unbounded.
+//! * [`StreamingEncoder`] — implements [`std::io::Write`] (`std` feature),
+//!   accepting bytes incrementally and flushing compressed output as blocks
+//!   fill. Requires `set_pledged_content_size` before the first write if
+//!   the Frame Content Size field is to be populated.
+//! * [`FrameCompressor`] — lower-level builder that owns the matcher and
+//!   the per-frame configuration; the streaming and one-shot helpers are
+//!   thin wrappers over it. Reach for it when you need to swap in a custom
+//!   [`Matcher`] implementation or share the matcher across frames.
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/mod.rs]: @coderabbitai Fixed in 1e737a7 — opening line now reads `Four entry points`.
